engine/subsystem/synchronize/sync_manager.go

```diff
@@ -0,0 +1,518 @@
+// synchronization. If an assets needs updating via REST it will push the work
+// to worker routines.
+func (m *Manager) controller() error {
+	// Pre-load all items for initial sync. The controller routine will take
+	// over this functionality to keep assets updated via REST only if needed.
+	wait, err := m.checkAllExchangeAssets()
+	if err != nil {
+		return err
+	}
+
+	if atomic.LoadInt32(&m.initSyncCompleted) != 1 {
+		log.Debugf(log.SyncMgr, "Exchange CurrencyPairSyncer initial sync started. %d items to process.",
+			m.createdCounter)
+		m.initSyncStartTime = time.Now()
+	}
+
+	go func(timer *time.Timer) {
+		for range timer.C {
+			if atomic.LoadInt32(&m.started) == 0 {
+				log.Debugln(log.SyncMgr, "Exchange CurrencyPairSyncer worker shutting down.")
+				return
+			}
+			wait, err = m.checkAllExchangeAssets()
+			if err != nil {
+				log.Errorf(log.SyncMgr, "Sync manager checking all exchange assets. Error: %s", err)
+				wait = m.getSmallestTimeout()
+			}
+			timer.Reset(wait)
+		}
+	}(time.NewTimer(wait))
+	return nil
+}
+
+// getSmallestTimeout returns the smallest configured timeout for all supported
+// synchronization protocols for controller sync.
+func (m *Manager) getSmallestTimeout() time.Duration {
+	if m.TimeoutREST < m.TimeoutWebsocket {
+		return m.TimeoutREST
+	}
+	return m.TimeoutWebsocket
+}
+
+// checkSyncItems checks agent against it's current last update time on all
+// individual synchronization items.
+func (m *Manager) checkSyncItems(exch exchange.IBotExchange, pair currency.Pair, a asset.Item, usingREST bool, update *time.Duration) {
+	agent := m.getAgent(exch.GetName(), pair, a, usingREST)
+	if m.SynchronizeOrderbook {
+		m.checkSyncItem(exch, &agent.Orderbook, agent, subsystem.Orderbook, update)
+	}
+	if m.SynchronizeTicker {
+		m.checkSyncItem(exch, &agent.Ticker, agent, subsystem.Ticker, update)
+	}
+	if m.SynchronizeTrades {
+		m.checkSyncItem(exch, &agent.Trade, agent, subsystem.Trade, update)
+	}
+}
+
+// checkSyncItem checks the individual sync item
+func (m *Manager) checkSyncItem(exch exchange.IBotExchange, indv *Base, agent *Agent, sync subsystem.SynchronizationType, update *time.Duration) {
+	until, needsUpdate := indv.NeedsUpdate(m.TimeoutREST, m.TimeoutWebsocket)
+	if needsUpdate {
+		// This needs to set processing in this controller routine so that it
+		// can be ignored on next check.
+		indv.SetProcessing(agent.Exchange, sync.String(), agent.Pair, agent.AssetType, m.TimeoutWebsocket, true)
+		m.sendJob(exch, agent.Pair, agent.AssetType, sync)
+	} else if until > 0 && until < *update {
+		*update = until
+	}
+}
+
+// sendJob sets agent base as processing for that ticker item then sends the
+// REST sync job to the jobs channel for processing.
+func (m *Manager) sendJob(exch exchange.IBotExchange, pair currency.Pair, a asset.Item, item subsystem.SynchronizationType) {
+	// NOTE: This is blocking, if there are no receivers and the buffer size is
+	// full then this will hang the controller.
```

Comment:
Yeah I will, at this point it becomes inconsequential anyway, it will be exceeding the buffer (10k) which is more pairs than what GCT supports (for now). I will put in a default case then log an error as something has gone terribly wrong.